components/actions_menu/actions_menu.tsx

```diff
@@ -0,0 +1,403 @@
+                        key={item.id + '_pluginmenuitem'}
+                        text={item.text}
+                        onClick={() => {
+                            if (item.action) {
+                                item.action(this.props.post.id);
+                            }
+                        }}
+                    />
+                );
+            }) || [];
+
+        let appBindings = [] as JSX.Element[];
+        if (this.props.appsEnabled && this.state.appBindings) {
+            appBindings = this.state.appBindings.map((item) => {
+                let icon: JSX.Element | undefined;
+                if (item.icon) {
+                    icon = (<img src={item.icon}/>);
+                }
+
+                return (
+                    <Menu.ItemAction
+                        text={item.label}
+                        key={item.app_id + item.location}
+                        onClick={() => this.onClickAppBinding(item)}
+                        icon={icon}
+                    />
+                );
+            });
+        }
+
+        const {formatMessage} = this.props.intl;
+
+        const divider = (
+            <li
+                className='MenuItem__divider'
+                role='menuitem'
+            />);
+
+        let marketPlace = null;
+        if (this.props.isSysAdmin) {
+            marketPlace = (
+                <>
+                    {divider}
+                    <Menu.ItemAction
+                        id={`marketplace_icon_${this.props.post.id}`}
+                        show={true}
+                        text={formatMessage({id: 'post_info.marketplace', defaultMessage: 'App Marketplace'})}
+                        icon={Utils.getMenuItemIcon('icon-view-grid-plus-outline')}
+                        onClick={this.handleOpenMarketplace}
+                    />
+                </>
+            );
+        }
+
+        if (isSystemMessage || !this.props.isSysAdmin) {
+            return null;
+        }
+
+        let menuItems;
+        let tutorialTip = null;
+        if (this.props.showTutorialTip) {
+            tutorialTip = (
+                <ActionsTutorialTip/>
+            );
+        }
+
+        const hasApps = appBindings.length > 0;
+        const hasPluggables = this.props.components[PLUGGABLE_COMPONENT] && this.props.components[PLUGGABLE_COMPONENT].length > 0;
+        const hasPluginItems = typeof pluginItems !== 'undefined' && pluginItems.length > 0;
+
+        if (hasPluginItems || hasApps || hasPluggables) {
+            const pluggable = (
+                <Pluggable
+                    postId={this.props.post.id}
+                    pluggableName={PLUGGABLE_COMPONENT}
+                />);
+
+            menuItems = [
+                pluginItems,
+                appBindings,
+                pluggable,
+                marketPlace,
+            ];
+        } else {
+            menuItems = [this.visitMarketplaceTip()];
+        }
+
+        return (
```

Comment:
Do we need any check or logic when menuItems length is 0?

Reply:
If there are no pluginItems, apps, or pluggable, menuItems will be set to this.visitMarketplace(). It will never be length === 0